### PR TITLE
add java support. Closes #4

### LINKED ISSUE #4: Java
Haven't tried it out yet but since they have native support it should be fine. Seems pretty convoluted compared to the others, I'm not familiar with Java though so I'd probably do something unconventional.

### PR DESCRIPTION
**Pull request by skarnecki**
Added java runtime support

### CODE REVIEW

**Comment by mthenw** [comment]
Hi mate!

File: plugins/java/java.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+// Open adds java defaults.
+func (p *Plugin) Open(fn *function.Function) error {
+	if fn.Runtime != Runtime {
+		return nil
+	}
+
+	fn.Runtime = RuntimeCanonical
+
+	if fn.Handler == "" {
+		fn.Handler = "example.Main::handle"
```
Comment (mthenw) [inline plugins/java/java.go]: example doesn't sounds good a a default package. maybe `main`? dunno :/

File: plugins/java/java_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+	"os"
+
+	"github.com/apex/apex/function"
+	"github.com/apex/log"
+	"github.com/jpillora/archive"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFunction_Build_ShouldUseGeneratedPom(t *testing.T) {
```
Comment (mthenw) [inline plugins/java/java_test.go]: check existing convention for test function names

File: _examples/java/README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
+# Requirements
```
Comment (mthenw) [inline _examples/java/README.md]: not sure where to put that directory. maybe docs?

File: plugins/java/java.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+	defer reader.Close()
+	if err != nil {
+		return err
+	}
+	for _, file := range reader.File {
+		fileReader, err := file.Open()
+		if err != nil {
+			return err
+		}
+		defer fileReader.Close()
```
Comment (tj) [inline plugins/java/java.go]: usually best to Close() at the end of the loop iteration so they're not left open

File: plugins/java/java.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+		fileContents, err := ioutil.ReadAll(fileReader)
+		if err != nil {
+			return err
+		}
+		zip.AddBytes(file.Name, fileContents)
+	}
+
+	fn.Log.Debug("cleaning mvn tmpfiles")
+	mvnCleanCmd := exec.Command("mvn", "clean")
+	mvnCleanCmd.Env = os.Environ()
```
Comment (tj) [inline plugins/java/java.go]: defaults to os.Environ()

File: plugins/java/java.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+		fn.Log.Debug("generating default pom")
+		generatedPom = true
+		if err := ioutil.WriteFile(expectedPomPath, []byte(genericPom), 0644); err != nil {
+			return err
+		}
+	}
+
+	fn.Log.Debug("creating jar")
+	mvnCmd := exec.Command("mvn", "package", "-Djar.finalName="+TargetJarFile)
+	mvnCmd.Env = os.Environ()
```
Comment (tj) [inline plugins/java/java.go]: this one too :D

**Comment by tj** [comment]
awesome thanks for tackling this! my java skills are non-existent

File: plugins/inference/inference.go [hunk 7]
```diff
@@ -17,6 +18,7 @@ func init() {
 			"main.py":  python.Runtime,
 			"index.js": nodejs.Runtime,
 			"main.go":  golang.Runtime,
+			"pom.xml":  java.Runtime,
```
Comment (mthenw) [inline plugins/inference/inference.go]: I think we can add another entry here: `"src/main/java/lambda/Main.java": java.Runtime` so default file can be handled

File: plugins/java/java.go [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+	"path/filepath"
+
+	"errors"
+
+	"github.com/apex/apex/function"
+	"github.com/jpillora/archive"
+)
+
+// prelude script template.
+var genericPom = `<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Comment (mthenw) [inline plugins/java/java.go]: could you move it to separate file

File: plugins/java/java.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+</project>
+`
+
+const (
+	// Runtime name used by Apex
+	Runtime = "java"
+	// RuntimeCanonical represents names used by AWS Lambda
+	RuntimeCanonical = "java8"
+	// TargetJarFile mvn target jar file name
+	TargetJarFile = "apex-plugin-target"
```
Comment (mthenw) [inline plugins/java/java.go]: no need to be public

File: _examples/java/functions/bar/function.json [hunk 10]
```diff
@@ -0,0 +1,10 @@
+{
+  "description": "Java example function",
+  "timeout": 1,
+  "memory": 256,
+  "runtime": "java",
+  "handler": "lambda.Example::handler",
+  "environment": {
```
Comment (mthenw) [inline _examples/java/functions/bar/function.json]: not used in this example

File: _examples/java/functions/bar/function.json [hunk 11]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (mthenw) [inline _examples/java/functions/bar/function.json]: actually this file is not needed

File: _examples/java/functions/foo/function.json [hunk 12]
```diff
@@ -0,0 +1,10 @@
+{
+  "description": "Java example function",
+  "timeout": 1,
+  "memory": 512,
+  "runtime": "java",
+  "handler": "lambda.Hello::handler",
+  "environment": {
```
Comment (mthenw) [inline _examples/java/functions/foo/function.json]: not needed

File: plugins/java/java.go [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+	fn.Log.Debug("creating jar")
+	mvnCmd := exec.Command("mvn", "package", "-Djar.finalName="+TargetJarFile)
+	mvnCmd.Dir = fn.Path
+	err := mvnCmd.Run()
+	if err != nil {
+		return err
+	}
+
+	expectedJarPath := filepath.Join(fn.Path, "/target/"+TargetJarFile+".jar")
```
Comment (mthenw) [inline plugins/java/java.go]: Join.(fn.Path, "target", TargetJarFile+".jar")

**Comment by mthenw** [comment]
LGTM, @tj please take a look

**Comment by tj** [comment]
few nitpicks but I'll tweak those up and amend! thanks man

**Pull request merged**